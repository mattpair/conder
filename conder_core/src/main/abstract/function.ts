import { AnySchemaInstance, AnyOpInstance, ow } from '../ops/index';
import { AnyNode, AnyRootNodeFromSet, BaseNodeDefs, RootNode } from "./IR";
import {MONGO_COMPILER, MONGO_GLOBAL_ABSTRACTION_REMOVAL} from './globals/mongo'
import { Transformer, Compiler } from "./compilers";


export type GlobalObject = {kind: "glob", name: string}

export type Manifest<F=FunctionDescription> = {
    globals: Map<string, GlobalObject>
    funcs: Map<string, F>
}

export type FunctionData<COMP=RootNode> = {
    readonly input: AnySchemaInstance[]
    readonly computation: COMP[]
}

export class FunctionDescription<COMP=RootNode> implements FunctionData<COMP>{
    public readonly input: AnySchemaInstance[]
    public readonly computation: COMP[]
    
    constructor(state: FunctionData<COMP>) {
        this.input = state.input
        this.computation = state.computation
    }

    public apply<NEW>(f: (c: COMP) => NEW[]): FunctionDescription<NEW> {
        return new FunctionDescription({
            input: this.input,
            computation: this.computation.flatMap(f)
        })
    }
}


export function toOps(funcs: Map<string, FunctionDescription>, override:  Compiler<RootNode> | undefined=undefined): Map<string, AnyOpInstance[]> {
    const ret: Map<string, AnyOpInstance[]> = new Map()
    const compiler: Compiler<RootNode> = override ? override : MONGO_GLOBAL_ABSTRACTION_REMOVAL.then(MONGO_COMPILER)
    // const computationLookup: Record<string, RootNode[]> = {}

    // ops.push(...compiler.run(funcs[k].computation))
    funcs.forEach((func, func_name) => {
        const ops: AnyOpInstance[] = [
            ow.assertHeapLen(func.input.length)
        ]
        func.input.forEach((schema, index) => {
            ops.push(
                ow.enforceSchemaInstanceOnHeap({heap_pos: index, schema}),
                ow.conditonallySkipXops(1),
                ow.raiseError("invalid input")
            )
        })
    
    
        ret.set(func_name, ops)
    })

    const compiled = compiler.run(funcs)
    funcs.forEach((func, func_name) => {
        ret.set(func_name, [...ret.get(func_name), ...compiled.get(func_name).computation])
    })
    return ret
}


export const OPSIFY_MANIFEST = new Transformer<
    Manifest, 
    Manifest<AnyOpInstance[]>>((man) => {
        const funcs = toOps(man.funcs)
        return {
            funcs,
            globals: man.globals
        }
    })