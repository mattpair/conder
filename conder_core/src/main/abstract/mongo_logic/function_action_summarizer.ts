import { apply, DummyVisitor } from './visitor';
import { PickTargetNode } from './../IR';
import { ActionSequence } from './lock_calculation';
import { MongoNodeSet } from '../globals/mongo';
import { TargetNodeSet, NodeSet } from "../IR";
import { ScopeMap } from './scope_map';

type ActionSummarizer = (n: TargetNodeSet<MongoNodeSet>[]) => ActionSequence

export const MONGO_ACTION_SUMMARIZER: ActionSummarizer = (nodes) => {
    apply(nodes, new DummyVisitor())
    return []
}

type GlobalTaints = Set<string>
