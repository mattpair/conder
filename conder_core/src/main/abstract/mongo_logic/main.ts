import { ActionSequence, calculate_lock_requirements, LockRequirements } from './lock_calculation';
import { MongoNodeSet } from '../globals/mongo';
import { TargetNodeSet, NodeSet } from "../IR";
import { ScopeMap } from '../../data_structures/scope_map';
import { MONGO_ACTION_SUMMARIZER } from './action_summarizer';

type LockCalculator = (input: Map<string, TargetNodeSet<MongoNodeSet>[]>) =>  Map<string, LockRequirements[string]>

export const MONGO_LOCK_CALCULATOR: LockCalculator = input => {
    
    const actions: Record<string, ActionSequence> = {}
    input.forEach((v, k) => {
        actions[k] = MONGO_ACTION_SUMMARIZER(v)
    })
    const locks = calculate_lock_requirements(actions)
    
    return new Map(Object.entries(locks))
}

