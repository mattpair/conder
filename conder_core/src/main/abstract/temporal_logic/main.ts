type Actions = {
    get: {id: string},
    mutation: {id: string, using: string[]},
}
type ActionKind = keyof Actions
type AnyAction = {
    [K in keyof Actions]: {kind: K} & Actions[K]
}[keyof Actions]


export type ActionSequence = AnyAction[]
export type LockRequirements = Record<string, {global: string, kind: "r" | "w"}[]>

export function calculate_lock_requirements(sequences: Record<string, ActionSequence>): LockRequirements {
    const actionAgainstData = new Map<string, Set<ActionKind>>()
    const lockReqs: LockRequirements = {}

    Object.values(sequences).forEach(seq => {
        seq.forEach(action => {
            const ops = actionAgainstData.get(action.id)
            if (ops === undefined) {
                actionAgainstData.set(action.id, new Set<ActionKind>([action.kind]))
            } else {
                ops.add(action.kind)
                actionAgainstData.set(action.id, ops)
            }
        })
    })
    
    Object.keys(sequences).forEach(func => {
        lockReqs[func] = []
        const previouslyGot = new Set<string>()
        sequences[func].forEach(action => {
            
            switch (action.kind) {
                case "get":
                    if (previouslyGot.has(action.id)) {
                        if (actionAgainstData.get(action.id).has("mutation")) {
                            lockReqs[func].push({global: action.id, kind: "r"})
                        }
                    } else {
                        previouslyGot.add(action.id)
                    }
                case "mutation":
                    break
                default: 
                    const n: never = action
            }
        })
    })

    return lockReqs
}