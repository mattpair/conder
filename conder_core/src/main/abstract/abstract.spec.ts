import { MONGO_COMPILER, MONGO_GLOBAL_ABSTRACTION_REMOVAL } from './globals/mongo';

import {Test, schemaFactory, AnyOpInstance} from '../ops/index'
import { AnyNode, RootNode } from 'src/main/abstract/IR'
import {BaseNodeDefs, PickNode, toOps, FunctionDescription, RootNodeCompiler } from '../../../index'
import { MONGO_UNPROVIDED_LOCK_CALCULATOR } from './mongo_logic/main';

type DagServer = Record<string, (...arg: any[]) => Promise<any>>
const TEST_STORE = "test"
const testCompiler: RootNodeCompiler =  MONGO_GLOBAL_ABSTRACTION_REMOVAL
    .tap((nonAbstractRepresentation) => {
        const locks = MONGO_UNPROVIDED_LOCK_CALCULATOR(nonAbstractRepresentation)
        expect(locks).toMatchSnapshot(`Required locks`)
    })
    .then(MONGO_COMPILER)

function withInputHarness(
    maybeStorage: "requires storage" | "no storage",
    proc_nodes: Record<string, FunctionDescription>,
    test: (server: DagServer) => Promise<void>): jest.ProvidesCallback {
    const getStoresAndProcedures = () => {
        const compiled = toOps(new Map(Object.entries(proc_nodes)), testCompiler)
        const PROCEDURES: Record<string, AnyOpInstance[]> = Object.fromEntries(compiled.entries())
        const STORES = {TEST_STORE: schemaFactory.Object({})}
        return {PROCEDURES, STORES}
    }
    if (maybeStorage === "requires storage") {
        return (cb) => {
            const spec = getStoresAndProcedures()
            return Test.Mongo.start({STORES: spec.STORES})
            .then(mongo => Test.Server.start({
                MONGO_CONNECTION_URI: `mongodb://localhost:${mongo.port}`,
                SCHEMAS: [],
                DEPLOYMENT_NAME: "statefultest",
                ...spec
            })
            .then(async server => {
                const testSurface: DagServer = {}
                for (const key in spec.PROCEDURES) {
                    testSurface[key] = (...args) => server.invoke(key, ...args)
                }
                return test(testSurface).then(() => {
                    server.kill()
                    mongo.kill()
                    cb()
                }).catch((e) => {
                    server.kill()
                    mongo.kill()
                    throw e
                })
            })
            
        )
        }
    }
    return (cb) => {
                            
        const spec = getStoresAndProcedures()
        return Test.Server.start({
            SCHEMAS: [],
            DEPLOYMENT_NAME: "statefultest",
            ...spec
        })
        .then(async server => {
            const testSurface: DagServer = {}
            for (const key in spec.PROCEDURES) {
                testSurface[key] = (...args) => server.invoke(key, ...args)
            }
            return test(testSurface).then(() => {
                server.kill()
                cb()
            }).catch((e) => {
                server.kill()
                throw e
            })
        })
    }
}

    

function noInputHarness(
    proc_nodes: Record<string, RootNode[]>, 
    test: (server: DagServer) => Promise<void>,
    maybeStorage: Parameters<typeof withInputHarness>[0]="no storage"
    ): jest.ProvidesCallback {
    const PROCEDURES: Record<string, FunctionDescription> = {}
    for (const key in proc_nodes) {
        PROCEDURES[key] = {
            input: [],
            computation: proc_nodes[key]
        }
    }
    
    return withInputHarness(maybeStorage,PROCEDURES, test)
}



describe("basic functionality", () => {
    

    it("return node returns null", 
        noInputHarness({
            r: [{kind: "Return"}]
        },
        async (server) => {
            const res = await server.r()
            expect(res).toBeNull()
        })
    )

    it("return node with value returns value",
        noInputHarness({
            r: [{
                kind: "Return", 
                value: {
                    kind: "Object", 
                    fields: [{
                        kind: "SetField", 
                        field_name: [{kind: "String", value: "some_field"}], 
                        value: {
                            kind: "Bool", 
                            value: false
                        }
                    }
                ]}
            }]
        }, async (server) => {
            expect(await server.r()).toEqual({some_field: false})
        })
    )

    it("can set double nested field",
        noInputHarness({
            r: [{
                    kind: "Save", 
                    index: 0,
                    value: {
                        kind: "Object", 
                        fields: [{
                            kind: "SetField", 
                            field_name: [{kind: "String", value: "nested"}], 
                            value: {
                                kind: "Object", 
                                fields: []
                            }
                        }]
                    }
                },
                {
                    kind: "Update",
                    target: {kind: "Saved", index: 0},
                    operation: {
                        kind: "SetField",
                        field_name: [{kind: "String", value: "nested"}, {kind: "String", value: "inside"}],
                        value: { kind: "String", value: "hello world"}
                    }
                },
                {kind: "Return", value: {kind: "Saved", index: 0}}
            ]
        }, async (server) => {
            expect(await server.r()).toEqual({nested: {inside: "hello world"}})
        })
    )

    it("allows deleting of fields on local objects",
        withInputHarness(
            "no storage", 
            {
                delete: {
                    input: [schemaFactory.Any], 
                    computation: [
                        {
                            kind: "Update", 
                            target: {kind: "Saved", index: 0},
                            operation: {kind: "DeleteField", field_name: [{kind: "String", value: "some_key"}]}
                        },
                        {
                            kind: "Return",
                            value: {kind: "Saved", index: 0}
                        }
                    ]}},
            async server => {
                expect(await server.delete({some_key: false, other: true})).toEqual({other: true})
                expect(await server.delete({})).toEqual({})
            }
        )
    )
    it("allows deleting of nested fields on locals",
        withInputHarness(
            "no storage", 
            {
                delete: {
                    input: [schemaFactory.Any], 
                    computation: [
                        {
                            kind: "Update", 
                            target: {kind: "Saved", index: 0},
                            operation: {
                                kind: "DeleteField", 
                                field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}]}
                        },
                        {
                            kind: "Return",
                            value: {kind: "Saved", index: 0}
                        }
                    ]}},
            async server => {
                expect(await server.delete({l1: {l2: false, other: true}})).toEqual({l1: {other: true}})
                expect(await server.delete({l1: {}})).toEqual({l1: {}})
            }
        )
    )

    it("can get nested field",
        noInputHarness({
            r: [{
                    kind: "Save", 
                    index: 0,
                    value: {
                        kind: "Object", 
                        fields: [{
                            kind: "SetField", 
                            field_name: [{kind: "String", value: "l1"}], 
                            value: {
                                kind: "Object", 
                                fields: []
                            }
                        }]
                    }
                },
                {
                    kind: "Update",
                    target: {kind: "Saved", index: 0},
                    operation: {
                        kind: "SetField",
                        field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}],
                        value: { kind: "String", value: "hello world"}
                    }
                },
                {kind: "Return", value: {
                    kind: "GetField", 
                    field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}],
                    target: {kind: "Saved", index: 0}
                }}
            ]
        }, async (server) => {
            expect(await server.r()).toEqual("hello world")
        })
    )

    function nComp(sign: PickNode<"Comparison">["sign"]): RootNode[] {
        return [{
            kind: "Return",
            value: {
                kind: "Comparison",
                sign,
                left: {kind: "Int", value: 1},
                right: {kind: "Int", value: 1}
            }
        }]
    }
    it("can compare numbers", 
        noInputHarness({
            geq: nComp(">="),
            leq: nComp("<="),
            l: nComp("<"),
            g: nComp(">"),
            e: nComp("=="),
            ne: nComp("!="),
        }, async server => {
            expect(await server.leq()).toBeTruthy()
            expect(await server.geq()).toBeTruthy()
            expect(await server.l()).toBeFalsy()
            expect(await server.g()).toBeFalsy()
            expect(await server.e()).toBeTruthy()
            expect(await server.ne()).toBeFalsy()
        })
    )

    function boolAlgTest(sign: PickNode<"BoolAlg">["sign"], left: PickNode<"BoolAlg">["left"], right: PickNode<"BoolAlg">["right"]): RootNode[] {
        return [{
            kind: "Return",
            value: {
                kind: "BoolAlg",
                left,
                right,
                sign
            }
        }]
    }
    it("can handle boolean algebra", 
        noInputHarness({
            trueNtrue: boolAlgTest("and", {kind: "Bool", value: true}, {kind: "Bool", value: true}),
            falseNtrue: boolAlgTest("and", {kind: "Bool", value: false}, {kind: "Bool", value: true}),
            trueNfalse: boolAlgTest("and", {kind: "Bool", value: true}, {kind: "Bool", value: false}),
            trueOtrue: boolAlgTest("or", {kind: "Bool", value: true}, {kind: "Bool", value: true}),
            falseOtrue: boolAlgTest("or", {kind: "Bool", value: false}, {kind: "Bool", value: true}),
            trueOfalse: boolAlgTest("or", {kind: "Bool", value: true}, {kind: "Bool", value: false}),
            falseOfalse: boolAlgTest("or", {kind: "Bool", value: false}, {kind: "Bool", value: false})

        }, async server => {
            expect(await server.trueNtrue()).toBeTruthy()
            expect(await server.falseNtrue()).toBeFalsy()
            expect(await server.trueNfalse()).toBeFalsy()
            expect(await server.trueOtrue()).toBeTruthy()
            expect(await server.trueOfalse()).toBeTruthy()
            expect(await server.falseOtrue()).toBeTruthy()
            expect(await server.falseOfalse()).toBeFalsy()
        })
    )

    it("supports basic math",
        noInputHarness({
            minus: [{kind: "Return", value: {
                kind: "Math",
                left: {kind: "Int", value: 42},
                right: {kind: "Int", value: -42},
                sign: "-"
            }}]
        }, async server => {
            expect(await server.minus()).toBe(84)
        })
    
    )

    it("allows if statements", 
        noInputHarness({
            ifTrue: [{
                kind: "If",
                cond: {kind: "Bool", value: true},
                ifTrue: {kind: "Return", value: {kind: "Int", value: 1}}
            }],
            ifFalseNoFinally: [{
                kind: "If",
                cond: {kind: "Bool", value: false},
                ifTrue: {kind: "Return", value: {kind: "Int", value: 1}}
            }],
            ifFalseFinally: [{
                kind: "If",
                cond: {kind: "Bool", value: false},
                ifTrue: {kind: "Return", value: {kind: "Int", value: 1}},
                finally: {kind: "Return", value: {kind: "Int", value: 2}}
            }]
        }, 
            async server => {
                expect(await server.ifTrue()).toBe(1)
                expect(await server.ifFalseNoFinally()).toBeNull()
                expect(await server.ifFalseFinally()).toBe(2)
        })
    )
})

describe("with input", () => {
    it("validates input", withInputHarness("no storage",{
        accepts3any: {
            input: [
                schemaFactory.Any,
                schemaFactory.Any,
                schemaFactory.Any
            ],
            computation: [{
                kind: "Return",
                value: {
                    kind: "Saved",
                    index: 2
                }
            }]
        }
    }, async server => {
        await expect(server.accepts3any("a", "b", "c", "d")).rejects.toThrowError()
        await expect(server.accepts3any("a", "b",)).rejects.toThrowError()
        expect(await server.accepts3any("a", "b", "c")).toEqual("c")
    }))

    it("check if field exists", withInputHarness("no storage",{
        checksField: {
            input: [
                schemaFactory.Any
            ],
            computation: [{
                kind: "Return",
                value: {
                    kind: "FieldExists",
                    field: {kind: "String", value: "test"},
                    value: {kind: "Saved", index: 0}
                }
            }]
        }
    }, async server => {
        expect(await server.checksField({test: "some"})).toBeTruthy()
        expect(await server.checksField({test: null})).toBeFalsy()
        expect(await server.checksField({t: "a"})).toBeFalsy()
    }))
    
})

describe("global objects", () => {

    const get: RootNode[] = [
        {
            kind: "Return", 
            value: {
                kind: "GetField", 
                target: {
                    kind: "GlobalObject", 
                    name: TEST_STORE
                },
                field_name: [{
                    kind: "String",
                    value: "l1"
                }]
            }
        }
    ]
    it("getting a non existent key returns null", 
        noInputHarness({
            get,
        }, 
        async server => {
            expect(await server.get()).toBeNull()
        }, 
        "requires storage")
    )

    const set: RootNode[] = [{
        kind: "Update",
        target: {kind: "GlobalObject", name: TEST_STORE},
        operation: {
            kind: "SetField",
            field_name: [{kind: "String", value: "l1"}],
            value: {kind: "Object", fields: [
                {
                    kind: "SetField", 
                    value: {
                        kind: "Int", value: 42
                    },
                    field_name: [{kind: "String", value: "l2"}]
                }
            ]}
        }
    }]

    it("getting a key returns the value",
        noInputHarness({
            get,
            set
        },
        async server => {
            expect(await server.set()).toBeNull()
            expect(await server.get()).toEqual({l2: 42})

        },
        "requires storage"
        )
    )

    const getNested: RootNode[] = [
        {
            kind: "Return", 
            value: {
                kind: "GetField", 
                target: {
                    kind: "GlobalObject", 
                    name: TEST_STORE
                },
                field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}]
            }
        }
    ]
  
    it("getting a non existent nested field throws an error",
        noInputHarness({getNested},
        async server => {
            await expect(server.getNested()).rejects.toThrowError()
        },
        "requires storage"
        )
    )

    const setNested: RootNode[] = [{
        kind: "Update",
        target: {kind: "GlobalObject", name: TEST_STORE},
        operation: {
            kind: "SetField",
            field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}],
            value: {
                kind: "Int", value: 41
            }             
        }
    }]
    it("setting a nested key on a non existent object throws an error",
        noInputHarness(
            {
                setNested,
                get,
                getNested
            },
            async server => {
                await expect(server.setNested()).rejects.toThrowError()
                await expect(server.getNested()).rejects.toThrowError()
                expect(await server.get()).toBeNull()
            },
            "requires storage"
        )
    )

    it("setting a nested key on an existing object",
        noInputHarness(
            {
                setNested,
                set,
                get,
                getNested
            },
            async server => {
                expect(await server.set()).toBeNull()
                expect(await server.setNested()).toBeNull()
                expect(await server.getNested()).toEqual(41)
                expect(await server.get()).toEqual({l2: 41})
                expect(await server.set()).toBeNull()
                expect(await server.getNested()).toEqual(42)
            },
            "requires storage"
        )
    )

    const checkL1: RootNode[] = [{
        kind: "Return",
        value: {
            kind: "FieldExists",
            value: {kind: "GlobalObject", name: TEST_STORE},
            field: {kind: "String", value: "l1"}
        }
    }]
    it("can check existence of keys",
        noInputHarness(
            {
                set,
                checkL1
            },
            async server => {
                expect(await server.checkL1()).toBe(false)
                expect(await server.set()).toBeNull()
                expect(await server.checkL1()).toBe(true)
            },
            "requires storage"
        )
    )


    it("allows deleting of keys on global objects",
        noInputHarness(
            
            {
                delete: [
                        {
                            kind: "Update", 
                            target: {kind: "GlobalObject", name: TEST_STORE},
                            operation: {kind: "DeleteField", field_name: [{kind: "String", value: "l1"}]}
                        }
                    ],
                set,
                get
            },
            async server => {
                expect(await server.delete()).toBeNull()
                expect(await server.set()).toBeNull()
                expect(await server.get()).toEqual({l2: 42})
                expect(await server.delete()).toBeNull()
                expect(await server.get()).toBeNull()
            },
            "requires storage"
        )
    )

    it("allows deleting of nested fields on globals",
        noInputHarness( 
            {
                delete: [
                        {
                            kind: "Update", 
                            target: {kind: "GlobalObject", name: TEST_STORE},
                            operation: {
                                kind: "DeleteField", 
                                field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}]}
                        }
                    ],
                set,
                get,
            },
            async server => {
                // Deleting a nested field on an object that does not exist is acceptable.
                expect(await server.delete()).toBeNull()
                expect(await server.set()).toBeNull()
                expect(await server.get()).toEqual({l2: 42})
                expect(await server.delete()).toBeNull()
                expect(await server.get()).toEqual({})
            },
            "requires storage"
        )
    )

    describe("race condition possible actions", () => {
        it("can perform updates that depend on global state", noInputHarness(
            {
                get, 
                set,
                setToSelfPlusOne: [{
                    kind: "Update",
                    target: {kind: "GlobalObject", name: TEST_STORE},
                    operation: {
                        kind: "SetField",
                        field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}],
                        value: {
                            kind: "Math",
                            left: {
                                kind: "GetField", 
                                target: {kind: "GlobalObject", name: TEST_STORE},
                                field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}]
                            },
                            right: {kind: "Int", value: 1},
                            sign: "+"
                        }
                    }
                }]
            },
            async server => {
                expect(await server.set()).toBeNull()
                expect(await server.setToSelfPlusOne()).toBeNull()
                expect(await server.get()).toEqual({l2: 43})
            }, "requires storage")
        )

        it("can perform updates that depend on some other global state", noInputHarness(
            {
                get, 
                setOther: [{
                    kind: "Update", 
                    target: {kind: "GlobalObject", name: "other"}, 
                    operation: {
                        kind: "SetField", 
                        field_name: [{kind: "String", value: "l1"}],
                        value: {kind: "Int", value: 734}
                    }
                }],
                setToOtherPlusOne: [{
                    kind: "Update",
                    target: {kind: "GlobalObject", name: TEST_STORE},
                    operation: {
                        kind: "SetField",
                        field_name: [{kind: "String", value: "l1"}],
                        value: {
                            kind: "Math",
                            left: {
                                kind: "GetField", 
                                target: {kind: "GlobalObject", name: "other"},
                                field_name: [{kind: "String", value: "l1"}]
                            },
                            right: {kind: "Int", value: 1},
                            sign: "+"
                        }
                    }
                }]
            },
            async server => {
                expect(await server.setOther()).toBeNull()
                expect(await server.setToOtherPlusOne()).toBeNull()
                expect(await server.get()).toBe(735)
            }, "requires storage")
        )


        it("can perform updates that depend on global state transitively", noInputHarness(
            {
                get, 
                set,
                setToSelfPlusOne: [
                    {
                        kind: "Save", index: 0,
                        value: {
                            kind: "GetField", 
                            target: {kind: "GlobalObject", name: TEST_STORE},
                            field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}]
                        }
                    },
                    {
                    kind: "Update",
                    target: {kind: "GlobalObject", name: TEST_STORE},
                    operation: {
                        kind: "SetField",
                        field_name: [{kind: "String", value: "l1"}, {kind: "String", value: "l2"}],
                        value: {
                            kind: "Math",
                            left: {kind: "Saved", index: 0},
                            right: {kind: "Int", value: 1},
                            sign: "+"
                        }
                    }
                }]
            },
            async server => {
                expect(await server.set()).toBeNull()
                expect(await server.setToSelfPlusOne()).toBeNull()
                expect(await server.get()).toEqual({l2: 43})
            }, "requires storage")
        )
    })

})