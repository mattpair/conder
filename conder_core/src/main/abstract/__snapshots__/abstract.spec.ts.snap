// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality allows deleting of fields on local objects: Required locks 1`] = `
Map {
  "delete" => Map {},
}
`;

exports[`basic functionality allows deleting of nested fields on locals: Required locks 1`] = `
Map {
  "delete" => Map {},
}
`;

exports[`basic functionality allows directly indexing into object keys: Required locks 1`] = `
Map {
  "getKeys" => Map {},
}
`;

exports[`basic functionality allows else ifs: Required locks 1`] = `
Map {
  "elseIfs" => Map {},
}
`;

exports[`basic functionality allows elses: Required locks 1`] = `
Map {
  "else" => Map {},
}
`;

exports[`basic functionality allows gathering of keys on objects: Required locks 1`] = `
Map {
  "getKeys" => Map {},
}
`;

exports[`basic functionality allows if statements: Required locks 1`] = `
Map {
  "ifTrue" => Map {},
  "ifFalseNoFinally" => Map {},
  "ifFalseFinally" => Map {},
}
`;

exports[`basic functionality allows indexing into arrays with an int: Required locks 1`] = `
Map {
  "getFirst" => Map {},
}
`;

exports[`basic functionality allows pushing to local arrays: Required locks 1`] = `
Map {
  "push" => Map {},
}
`;

exports[`basic functionality allows pushing to nested local arrays: Required locks 1`] = `
Map {
  "push" => Map {},
}
`;

exports[`basic functionality can compare numbers: Required locks 1`] = `
Map {
  "geq" => Map {},
  "leq" => Map {},
  "l" => Map {},
  "g" => Map {},
  "e" => Map {},
  "ne" => Map {},
}
`;

exports[`basic functionality can get nested field: Required locks 1`] = `
Map {
  "r" => Map {},
}
`;

exports[`basic functionality can handle boolean algebra: Required locks 1`] = `
Map {
  "trueNtrue" => Map {},
  "falseNtrue" => Map {},
  "trueNfalse" => Map {},
  "trueOtrue" => Map {},
  "falseOtrue" => Map {},
  "trueOfalse" => Map {},
  "falseOfalse" => Map {},
}
`;

exports[`basic functionality can set double nested field: Required locks 1`] = `
Map {
  "r" => Map {},
}
`;

exports[`basic functionality cleans up after for eachs: Required locks 1`] = `
Map {
  "forVars" => Map {},
}
`;

exports[`basic functionality cleans up after ifs: Required locks 1`] = `
Map {
  "ifVars" => Map {},
}
`;

exports[`basic functionality return node returns null: Required locks 1`] = `
Map {
  "r" => Map {},
}
`;

exports[`basic functionality return node with value returns value: Required locks 1`] = `
Map {
  "r" => Map {},
}
`;

exports[`basic functionality supports basic math: Required locks 1`] = `
Map {
  "minus" => Map {},
}
`;

exports[`global objects Locks locks prevent progress if not held: Required locks 1`] = `
Map {
  "unsafeGet" => Map {},
  "unsafeSet" => Map {},
  "incr" => Map {},
}
`;

exports[`global objects allows calling functions: Required locks 1`] = `
Map {
  "caller" => Map {},
  "callee" => Map {},
}
`;

exports[`global objects allows conditional updates: Required locks 1`] = `
Map {
  "maybeSet" => Map {},
}
`;

exports[`global objects allows deleting of keys on global objects: Required locks 1`] = `
Map {
  "delete" => Map {},
  "set" => Map {},
  "get" => Map {},
}
`;

exports[`global objects allows deleting of nested fields on globals: Required locks 1`] = `
Map {
  "delete" => Map {},
  "set" => Map {},
  "get" => Map {},
}
`;

exports[`global objects allows getting a key with a number key: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
}
`;

exports[`global objects can check existence of keys: Required locks 1`] = `
Map {
  "set" => Map {},
  "checkL1" => Map {},
}
`;

exports[`global objects can get keys from global objects: Required locks 1`] = `
Map {
  "getKeys" => Map {},
  "setKeys" => Map {},
}
`;

exports[`global objects can perform updates to objects within arrays: Required locks 1`] = `
Map {
  "init" => Map {},
  "update" => Map {},
}
`;

exports[`global objects can push to arrays in global objects: Required locks 1`] = `
Map {
  "init" => Map {},
  "push" => Map {},
  "get" => Map {},
}
`;

exports[`global objects getting a key returns the value: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "getWhole" => Map {},
}
`;

exports[`global objects getting a non existent key returns null: Required locks 1`] = `
Map {
  "get" => Map {},
}
`;

exports[`global objects getting a non existent nested field throws an error: Required locks 1`] = `
Map {
  "getNested" => Map {},
}
`;

exports[`global objects race condition possible actions can perform updates that depend on global state - ifs: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setTo0If42" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions can perform updates that depend on global state transitively: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setToSelfPlusOne" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions can perform updates that depend on global state: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setToSelfPlusOne" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions can perform updates that depend on some other global state: Required locks 1`] = `
Map {
  "get" => Map {},
  "setOther" => Map {},
  "setToOtherPlusOne" => Map {
    "other" => "r",
  },
}
`;

exports[`global objects race condition possible actions global state taint is applied on partial updates to variables: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "updateWithPartialState" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions global state taint is applied on updates to variables: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setToSelfPlusOne" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions global state taint is erased on overwrites: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "updateWithOverwrittenState" => Map {},
}
`;

exports[`global objects race condition possible actions global state taint is transitive through variables: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setToSelfPlusOne" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions iterating over an array of globals while writing requires a lock: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setLookupKeys" => Map {},
  "deleteLookupFields" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions mutation conditional on same global state requires lock: Required locks 1`] = `
Map {
  "get" => Map {},
  "set" => Map {},
  "setTo0If42" => Map {
    "test" => "w",
  },
}
`;

exports[`global objects race condition possible actions pushing then returning: Required locks 1`] = `
Map {
  "push" => Map {},
}
`;

exports[`global objects race condition possible actions should allow checks of existence with comparisons to none: Required locks 1`] = `
Map {
  "get_user" => Map {},
  "create_user" => Map {
    "users" => "w",
  },
}
`;

exports[`global objects setting a nested key on a non existent object throws an error: Required locks 1`] = `
Map {
  "setNested" => Map {},
  "get" => Map {},
  "getNested" => Map {},
}
`;

exports[`global objects setting a nested key on an existing object: Required locks 1`] = `
Map {
  "setNested" => Map {},
  "set" => Map {},
  "get" => Map {},
  "getNested" => Map {},
}
`;

exports[`global objects strings allows concatenation with some types: Required locks 1`] = `
Map {
  "add" => Map {},
}
`;

exports[`roles allows use of stateful roles: Required locks 1`] = `
Map {
  "usersOnly" => Map {},
  "getUser" => Map {},
}
`;

exports[`roles functions can be guarded by roles: Required locks 1`] = `
Map {
  "adminsOnly" => Map {},
  "getAdminId" => Map {},
}
`;

exports[`with input check if field exists: Required locks 1`] = `
Map {
  "checksField" => Map {},
}
`;

exports[`with input validates input: Required locks 1`] = `
Map {
  "accepts3any" => Map {},
}
`;
